### ISSUE TITLE
Vulkan shows a black screen

### ISSUE THREAD

**Issue by BattleAxeVR** [NVIDIAGameWorks/NRDSample#3]
Hi, just tried compiling / running the latest sample, but Vulkan mode doesn't seem to be working properly. 

I'm using latest Nvidia non-beta drivers on my 3090, on Windows 11 Pro for Workstation w/ all latest Windows Updates, if that matters.

I tried all the scenes. Zero Day (scene 4) doesn't even start on Vulkan, but they all do on D3D12.

![image](https://user-images.githubusercontent.com/11604039/153128862-5d55a689-0159-44f9-9fe6-17bd11c68217.png)

**Comment by dzhdanNV** [comment]
It's not a bug :) If you open any RT shader you will see this:
```
#if( !defined( COMPILER_FXC ) && !defined( VULKAN ) )

// main code

#else

[numthreads( 16, 16, 1 )]
void main( uint2 pixelPos : SV_DispatchThreadId )
{
    // no TraceRayInline support, because of:
    //  - DXBC
    //  - SPIRV generation is blocked by https://github.com/microsoft/DirectXShaderCompiler/issues/4221
}

#endif
```
Sorry about it, but I didn't expect that DXC wouldn't be able to produce SPIRV for shaders containing `TraceRayInline`. SPIRV generation is blocked by this bug: https://github.com/microsoft/DirectXShaderCompiler/issues/4221 (created by a colleague, attached files are from NRD sample)

**Comment by BattleAxeVR** [comment]
Would it be too much to ask to fix this instead of waiting for official support upstream?

I can't update my NRD integration until I confirm it works 100% on Vulkan, and I don't have time to fix issues in sample code myself.

RTXDI sample uses the USE_RAY_QUERY macro / preprocessor define to toggle between raygen / queries, and as far as I remember, it works with Ray Query in Vulkan mode too.

#if USE_RAY_QUERY
RayQuery<RAY_FLAG_SKIP_PROCEDURAL_PRIMITIVES> rayQuery;

    rayQuery.TraceRayInline(SceneBVH, rayFlags, instanceMask, ray);
    .....................
#else
TraceRay(SceneBVH, rayFlags, instanceMask, 0, 0, 0, ray, payload);
#endif

**Comment by BattleAxeVR** [comment]
![image](https://user-images.githubusercontent.com/11604039/153642871-8c457dec-0cf8-4abd-a375-e5773e35edfa.png)

**Comment by BattleAxeVR** [comment]
If this combination of flags for AnyHit don't compile with current DXC,

RayQuery<RAY_FLAG_SKIP_PROCEDURAL_PRIMITIVES | RAY_FLAG_ACCEPT_FIRST_HIT_AND_END_SEARCH>

These combinations of flags certainly do compile in the RTXDI sample:

![image](https://user-images.githubusercontent.com/11604039/153645938-2af673e1-32b5-4285-9945-00793da1f695.png)

I tried re-enabling the shaders in the sample real quick by enabling that code path for vulkan, and changing the flag, but got some other compile errors. 

Bottom line, if RTXDI can run Vulkan + RayQueries via HLSL shaders using DXC to compile them, so should NRDSample.

**Comment by dzhdanNV** [comment]
> Bottom line, if RTXDI can run Vulkan + RayQueries via HLSL shaders using DXC to compile them, so should NRDSample.

A wrong assumption. I have 3 shaders with `TraceRayInline`, 1 of them compiles. The order of operations plus "who-know-what-else" trigger an internal error inside DXC. So...

> Would it be too much to ask to fix this instead of waiting for official support upstream?

I just can't.

> I can't update my NRD integration until I confirm it works 100% on Vulkan, and I don't have time to fix issues in sample code myself.

IMO - fully functional DX12 is enough to inspect IQ, NRD does have valid SPIRV generated from same source code, so it will work out of the box. Instead of investing time into "sample fixing", you could just update and validate in your app, which is "closer" to you, your expectations and your needs, obviously.

**Comment by BattleAxeVR** [comment]
I copy-pasted the ray query code into my own HLSL shaders that are compiled via DXC.exe and they compile just fine:

![image](https://user-images.githubusercontent.com/11604039/153648483-a4821799-ee99-4f72-a3e8-be13c156e8a7.png)

**Comment by BattleAxeVR** [comment]
It's almost certainly a bug in the sample, or at least has a workaround. I'm not sure what it is, but my issue with NRD sample isn't in my shaders (or yours), it's in the CPP / Vulkan side which is why I need to run it side-by-side to debug what's going wrong for me since 2.10. I get a crash deep inside VkUpdateDescriptorSets which was running fine in 2.5.

I need Vulkan to work in the sample to proceed here. I understand if you don't have time but there is certainly a problem in the NRDSample repo code causing it to not compile. Same ray query flags and code compiles in my own build system and in RTXDI too.

**Comment by dzhdanNV** [comment]
Please, take shaders from the NRD sample (not 10 lines of code) and compile them outside of the sample. Since the issue is public, this statement `It's almost certainly a bug in the sample` requires proving. In my opinion, if DXC produces valid DXIL and by using same code & parameters compilation gets aborted with a nonsense error, it means that something is wrong on the compiler side.

> it's in the CPP / Vulkan side which is why I need to run it side-by-side to debug what's going wrong for me since 2.10. I get a crash deep inside VkUpdateDescriptorSets which was running fine in 2.5.

NRD sample worked fine with NRD v2.5-v2.10 in Vulkan mode on Windows.

You mentioned that you needed to validate that NRD worked in Vulkan mode first. As a suitable solution you can update NRD in the RTXDI sample (unfortunately, some code massaging will be needed, but should be pretty straightforward, I believe).

What I can do?
- push for a solution on the compiler side
- blindly shuffle the shader code waiting for a successful compilation

**Comment by BattleAxeVR** [comment]
All I would ask in the future, is to make sure the code sample works in Vulkan too, before making API changes to the bleeding edge, because code samples not working has consequences for people who rely on it.

**Comment by dzhdanNV** [comment]
@BattleAxeVR DXC from latest Vulkan SDK compiles shaders correctly. RT shaders are enabled again for Vulkan in the latest release.
